File: core/src/rules/validate-required.ts
// core/src/rules/validate-required.ts
//
// Layer 2 atomic rule: checks Google Rich Results required-property
// constraints. Two flavors:
//   - simple required:    every name in `required` must be present
//   - "at least one of":  for every group in `requiredOneOf`, at least
//                         one member must be present
//
// Pure function. No I/O. Returns Issue[] (possibly empty).
//
// "Present" means: the property key exists on the object AND its
// value is not undefined or null. (An empty string is considered
// present, matching how schema.org markup typically reads in the
// wild.)

import type { Issue } from "../types.js";
import {
  missingRequiredProperty,
  missingRequiredPropertyOneOf,
} from "../errors.js";

function isPresent(obj: Record<string, unknown>, key: string): boolean {
  const v = obj[key];
  return v !== undefined && v !== null;
}

export function validateRequired(
  obj: Record<string, unknown>,
  typeName: string,
  required: string[],
  requiredOneOf: string[][],
): Issue[] {
  const issues: Issue[] = [];

  // Simple required: each missing one emits its own Issue.
  for (const propertyName of required) {
    if (!isPresent(obj, propertyName)) {
      issues.push(missingRequiredProperty(typeName, propertyName));
    }
  }

  // "At least one of": for each group, fail if no member is present.
  for (const alternatives of requiredOneOf) {
    if (alternatives.length === 0) continue;
    const anyPresent = alternatives.some((name) => isPresent(obj, name));
    if (!anyPresent) {
      issues.push(missingRequiredPropertyOneOf(typeName, alternatives));
    }
  }

  return issues;
}

File: python/src/schema_audit/rules/validate_required.py
"""Layer 2 atomic rule: Google Rich Results required-property constraints.

Mirrors ``core/src/rules/validate-required.ts`` 1:1. Two flavors:

- **simple required**: every name in ``required`` must be present
- **"at least one of"**: for every group in ``requiredOneOf``, at least
  one member must be present

"Present" means the property key exists on the dict AND its value is
not ``None``. (An empty string is considered present, matching how
schema.org markup typically reads in the wild.)
"""

from __future__ import annotations

from ..errors import missing_required_property, missing_required_property_one_of
from ..types import Issue


def _is_present(obj: dict[str, object], key: str) -> bool:
    return key in obj and obj[key] is not None


def validate_required(
    obj: dict[str, object],
    type_name: str,
    required: list[str],
    required_one_of: list[list[str]],
) -> list[Issue]:
    issues: list[Issue] = []

    for property_name in required:
        if not _is_present(obj, property_name):
            issues.append(missing_required_property(type_name, property_name))

    for alternatives in required_one_of:
        if not alternatives:
            continue
        if not any(_is_present(obj, name) for name in alternatives):
            issues.append(missing_required_property_one_of(type_name, alternatives))

    return issues
